### PR TITLE
Blockwise Sparse Sage Attention

### PR DESCRIPTION
**Pull request by ksikiric**
Add block-wise sparse attention support to SageAttention v1 and v2 (MXFP4)

## Motivation

Recent video diffusion models (e.g. HunyuanVideo 1.5) are trained with spatially sparse attention, the model learns to attend only to structurally relevant token subsets rather than the full sequence. These models expose their sparsity pattern as a block-level attention mask at inference time. Without kernel-level support, serving these models still requires computing the full dense attention over long sequences (e.g. 75,600 tokens), leaving significant performance on the table. This PR adds block-wise sparse attention to SageAttention v1 and v2 (MXFP4) so we can directly exploit the sparsity patterns these models were trained with, skipping entire Q-KV block pairs and translating the model's learned sparsity into real latency and compute savings.

## Technical Details

**Block-wise sparse attention for SageAttention v1 (fav3_sage) and v2/MXFP4 (fav3_sage_mxfp4):** Adds a ragged block LUT mechanism that allows Q blocks to skip masked-out KV blocks entirely, avoiding unnecessary computation. The kernel is split into mask and no-mask subkernels for optimal performance.

**GPU-accelerated block LUT generation:** Introduces a Triton kernel (block_lut.py) to convert a dense block attention mask into a compact ragged LUT (kv_block_indices, lut_start, lut_count) on-device, avoiding host round-trips.

**Benchmark infrastructure for sparse attention:** Extends both bench_fav3_sage.py and bench_fav3_sage_mxfp4.py with --block_sparsity, --n_repetitions (throughput statistics over multiple random masks with median/Q1/Q3/p10/p90), --compare_to_ref, and CSV output (-o). Also adds attention_ref_block_sparse as a PyTorch reference for correctness validation.

## Test Plan

- [x] pytest op_tests/triton_tests/attention/test_fav3_sage.py

- [x] bench_fav3_sage.py with --block_sparsity 0.5 --n_repetitions 10

- [x] bench_fav3_sage_mxfp4.py with --block_sparsity 0.5 --n_repetitions 10

- [x] Dense (no sparsity) benchmarks remain unchanged in behavior

## Test Result

All tests in test_fav3_sage.py pass, including the new block-sparse correctness tests for both v1 and MXFP4 kernels (dense-vs-sparse equivalence, block-sparse-vs-reference, empty KV block handling, and backward compatibility with block_lut=None).

Benchmarks perform as expected: throughput scales with sparsity level (higher sparsity = fewer blocks computed = lower latency), and the --n_repetitions flow correctly reports stable statistics across multiple random masks. Dense benchmarks (no sparsity flags) remain unchanged in behavior.

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: aiter/ops/triton/attention/fav3_sage_attention_mxfp4_wrapper.py [hunk 1]
```diff
@@ -212,6 +239,18 @@ def fav3_sage_mxfp4_func(
     padded_d_qk = max(16, 1 << (head_size_qk - 1).bit_length())
     padded_d_v = max(16, 1 << (head_size_v - 1).bit_length())
 
+    # Block sparse logic
+    if use_block_sparse:
+        if kv_block_indices is None or lut_start is None or lut_count is None:
```
Comment (juuso-oskari) [inline aiter/ops/triton/attention/fav3_sage_attention_mxfp4_wrapper.py:250]: assert not causal here

File: aiter/ops/triton/attention/fav3_sage.py [hunk 2]
```diff
@@ -399,6 +410,17 @@ def fav3_sage_func(
     window_size_left, window_size_right = int(window_size[0]), int(window_size[1])
     use_sliding_window = window_size_left != -1 or window_size_right != -1
 
+    if use_block_sparse:
+        if kv_block_indices is None or lut_start is None or lut_count is None:
```
Comment (juuso-oskari) [inline aiter/ops/triton/attention/fav3_sage.py:420]: assert not causal here

**Review by juuso-oskari — COMMENTED** [review]
otherwise LGTM!

File: op_tests/op_benchmarks/triton/bench_fav3_sage.py [hunk 3]
```diff
@@ -801,6 +1370,38 @@ def main():
 
     if not args.dv:
         args.dv = args.d
+
+    # Block-sparse: restrict to fav3_sage and validate
+    block_sparse = args.block_sparsity is not None or getattr(
+        args, "block_mask_file", None
+    )
+    if block_sparse:
+        args.fav3_sage = True
```
[the remark this answers is not in the document]
Comment (ksikiric) [reply op_tests/op_benchmarks/triton/bench_fav3_sage.py:1379]: This check is done `make_block_attn_mask `called from inside: `run_benchmark_captured`, when `--load_captured` is aused. In the case neither `block_sparsity `nor `block_mask_file `is provided, dense sage will be called.
